libraries/stdlib/genglsl/mx_normal_map.glsl

```diff
@@ -1,11 +1,11 @@
-void mx_normalmap(vec3 value, int tangent_space, vec3 N, vec3 T, out vec3 result)
+void mx_normal_map(vec3 value, int map_space, float normal_scale, vec3 N, vec3 T,  out vec3 result)
```

Comment:
This is somewhat of an edge case, but historically we've omitted underscores for node names in the standard library (e.g. tiledimage), adding them only to nodes in new libraries such as pbrlib and bxdf.  Perhaps it would make sense to omit the underscore here, since this node now belongs to the standard library?

Reply:
This is the implementation, but I assume your referring to the to the nodedef name? We could keep the <normalmap>
name as it would also be backwards compatible. I think it was Doug who suggested that this be renamed. Whatever name you decide on we case use.

Reply:
I think I went along with the name "normal_map" as suggested by Niklas when it was added to the pbrlib.  If it's going to be moved to stdlib, I think it should be called "**normalmap**" and go into the "convolution2d" nodegroup next to "heighttonormal".  As an aside, the nodegroup name "convolution2d" is maybe not the best choice at this point, since neither heighttonormal or normalmap actually perform convolution, and the one node that does ("blur") is not implemented anywhere that I know of.

Reply:
I believe heighttonormal does perform a simple convolution, since it needs to detect gradients in the height field.  So it’s just normalmap that would be the odd man out.  Perhaps it’s well placed as a texture node?

Reply:
Except that as currently implemented, "normalmap" doesn't actually read from an image file- it instead has an "in" input that I would expect would be connected to an image node.  Here's what's currently in the pbrlib_defs.mtlx file (though I am noticing now that "tangent_space" should probably be a boolean not an integer):
  <!--
    Node: <normalmap>; will be moved to stdlib?
    Also expecting <heightmap> node?
  -->   
  <nodedef name="ND_normalmap" node="normalmap" type="vector3" nodegroup="pbr">
    <input name="in" type="vector3" value="0.5, 0.5, 1.0"/>
    <parameter name="tangent_space" type="integer" value="1"/>
    <input name="normal" type="vector3" defaultgeomprop="Nworld"/>
    <input name="tangent" type="vector3" defaultgeomprop="Tworld"/>
  </nodedef>

Reply:
>   <!--
>     Node: <normalmap>; will be moved to stdlib?
>     Also expecting <heightmap> node
>   -->   
>   <nodedef name="ND_normalmap" node="normalmap" type="vector3" nodegroup="pbr">
>     <input name="in" type="vector3" value="0.5, 0.5, 1.0"/>
>     <parameter name="tangent_space" type="integer" value="1"/>
>     <input name="normal" type="vector3" defaultgeomprop="Nworld"/>
>     <input name="tangent" type="vector3" defaultgeomprop="Tworld"/>
>   </nodedef>

Reply:
Well for some reason I can't paste the text or a file in here, so please forgive the lack of proper formatting:
nodedef name="ND_normalmap" node="normalmap" type="vector3"
input name="in" type="vector3" value="0.5, 0.5, 1.0"
parameter name="tangent_space" type="integer" value="1"
input name="normal" type="vector3" defaultgeomprop="Nworld"
input name="tangent" type="vector3" defaulteomprop="Tworld"

Reply:
Maybe it should be a "math" node, same as transformpoint/normal/vector?

Reply:
That works for me, and I’d be interested in thoughts from Bernard.

Reply:
Yes. it is simply a vector transform so "math" sounds good. I'll also keep the name as "normalmap" and move it after the place2d "math" definition as that seems to be closest image related item.

Reply:
Bernard- would you mind sending me a quick description of "normalmap" that I can put into the actual Spec document?

Reply:
Did you also switch "tangentspace" to be a boolean?  Would be helpful to get a one-line description of each input and parameter for the spec as well.
There was also talk of adding a "heightmap" node- anything on that?

Reply:
Let me know if this is sufficient:
--------------------------------
The <normalmap> node is a math node that allows for a normal vector defined in one space to be transformed into another space. Inputs:
- "in" : input normal vector
- "space" : the input space for the normal. Possible options are "object" for object space and "tangent" for tangent space.
- "scale" : the normal can be scaled (offset) by a given multiplier.
- "normal" and "tangent". Two vectors from which a bitangent vector is computed to form the basis used for transformation. By default the vectors are in world space, meaning the output vector will be in world space.
```
  <nodedef name="ND_normalmap" node="normalmap" type="vector3" nodegroup="math">
    <input name="in" type="vector3" value="0.5, 0.5, 1.0"/> 
    <parameter name="space" type="string" value="tangent" enum="object,tangent"/>
    <input name="scale" type="float" value="1.0"/>
    <input name="normal" type="vector3" defaultgeomprop="Nworld"/>
    <input name="tangent" type="vector3" defaultgeomprop="Tworld"/>
  </nodedef>
```
Below is an example of this utility being used to read in a image containing tangent space normals and outputting world space normals scaled by a factor of 1.5.
```
  <image name="normal_image" type="vector3" colorspace="Raw">
    <parameter name="file" type="filename" value="normal_image.png" />
  </image>
  <normalmap name="surface_normals" type="vector3">
    <input name="in" type="vector3" nodename="normal_image" />
    <input name="scale" type="float" value="1.5" />
  </normalmap>
```
--------------------------------

We did have the discussion about add in a <heightmap> node which would take file image as input, but have decided that the existing <heighttonormal> is sufficient though we may want to add another kernel option.